--- ff12_p1.c	Thu Apr 21 19:34:07 2016
+++ ff12_p2.c	Thu Apr 21 19:37:11 2016
@@ -5323,7 +5323,7 @@
 	if (n_vol < 0x10000) {					/* Number of total sectors */
 		st_word(tbl + BPB_TotSec16, (WORD)n_vol);
 	} else {
-		st_dword(tbl + BPB_TotSec32, (WORD)n_vol);
+		st_dword(tbl + BPB_TotSec32, n_vol);
 	}
 	tbl[BPB_Media] = md;					/* Media descriptor */
 	st_word(tbl + BPB_SecPerTrk, 63);		/* Number of sectors per track */
